### PR TITLE
[collectd 6] Add an option to configure resource attributes.

### PR DESCRIPTION
**Pull request by octo**
ChangeLog: daemon: The new "Resource" option allows users to configure resource attributes manually if desired.

### CODE REVIEW

File: src/collectd.conf.pod [hunk 1]
```diff
@@ -328,6 +328,51 @@ the daemon should try to figure out the "fully qualified domain name", FQDN.
 This is done using a lookup of the name returned by C<gethostname>. This option
 is enabled by default.
 
+=item B<Resource> I<Type>
+
+The B<Resource> block defined the I<Resource Attributes> to use for local
+metrics such as CPU usage, i.e. metrics that were not received from a remote
```
Comment (eero-t) [inline src/collectd.conf.pod:334]: Reads better to me when tenses are switched:
```suggestion
B<Resource> block defines I<Resource Attributes> to be used for local
metrics such as CPU usage, i.e. metrics that were not received from a remote
```

**Review by eero-t — CHANGES_REQUESTED** [review]
Code changes look fine.   But I think documentation needs some improvements:
* Name is too generic, so some short explanation how this "Resource" differs from all the other "Resource" items already mentioned in the document, would be good
* Text is on the abstract side and IMHO would need some example of how it changes collectd output, so users unfamiliar with OpenTelemetry know what it's about

**Comment by eero-t** [comment]
There should be also `None` option for disabling resource attribute labels completely, because resource label names use characters that are not valid for Prometheus metrics, see https://github.com/collectd/collectd/issues/4204 and: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

**Comment by eero-t** [comment]
> There should be also `None` option for disabling resource attribute labels completely, because resource label names use characters that are not valid for Prometheus metrics, see #4204 and: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

That issue is now fixed, but need for disabling the resource labels has gone nowhere, because they just duplicate information that Prometheus **already** adds to the metrics.

**Comment by octo** [comment]
My thinking was that the following config should give you the same effect as a `None` resource:

```
<Resource Generic>
  Attribute "service.name" ""
</Resource>
```

Can you tell me more how the resource attributes and the Prometheus export clash? It's important to me to get this right.

File: src/collectd.conf.pod (hunk 1 above)
Comment (octo) [reply src/collectd.conf.pod]: Fixed the typo. I prefer the active voice though.

**Comment by eero-t** [comment]
> My thinking was that the following config should give you the same effect as a `None` resource:

True, but I'm assuming that you're going to be adding more attributes in the future?

`None` option would be a more user-friendly way of making sure that all extra labels are disabled.

> Can you tell me more how the resource attributes and the Prometheus export clash? It's important to me to get this right.

Problem is that extra labels (on every metric) can noticeably increase Prometheus resource usage in larger clusters, so it's important that one can easily disable anything that is not needed, especially _when Prometheus already provides corresponding information_ in (more relevant kubernetes namespace) labels.

**Comment by octo** [comment]
> > My thinking was that the following config should give you the same effect as a `None` resource:
> 
> True, and it's a useful feature, but I'm assuming that you're going to be adding more attributes in the future?

My design idea is this:

* Metrics that already have resource attributes, e.g. because they were received from somewhere else via gRPC or something similar, are simply passed through. No resource attributes are added, removed, or modified.
* Metrics that are passed to `plugin_dispatch_metric_family` without resource attributes are considered to be "local", i.e. as originating from this collectd instance. The resource attributes associated with this instance of collectd are attached to the metric family.
* Strike a good balance between auto-discovery and user control. Users likely don't want to manually configure the host name. If you're managing a large fleet, having a different config file for each host is quite annoying.
    * Provide presets that users can chose by passing an appropriate argument to the `Preset` block. At the moment, `Generic` and `Host` are the only available presets. Other presets, e.g. for GCE instances, K8s pods, etc. might follow.
    * Allow users to amend or overwrite the defaults, using the `Attribute` option inside the `Resource` block.

> `None` option would be a more user-friendly way of making sure that _all_ extra labels are disabled.

I think you're coming from a position where these labels cause problems with the *write_prometheus plugin*. Once these are properly integrated (I'm working on this in #4213) I think that the existing `Generic` preset provides more utility for you than a `None` preset, because they allow you to set the `job` and `instance` labels appropriately. 

> Problem is that extra labels (on every metric) can noticeably increase Prometheus resource usage in larger clusters, so it's important that one can easily disable anything that is not needed, especially _when Prometheus already provides corresponding information_ in (more relevant kubernetes namespace) labels.

Gotcha, thanks for the info! IMHO you have two options already:

1. Use the `Generic` preset and unset the one resource attribute it installs if required, as outlined above.
2. Use metric relabeling in Prometheus to discard unwanted labels.

I'm sure collectd will gain additional features in the future, but we have to start somewhere, and this seems like quite a good position to be in for a first step.

**Review by eero-t — APPROVED** [review]
Ok, seems reasonable, especially considering https://github.com/collectd/collectd/pull/4213.

**Pull request merged**